tests/codex/testasyncstreamwrapper.nim

```diff
@@ -0,0 +1,90 @@
+import pkg/asynctest
+import pkg/chronos
+import pkg/chronos/transports/stream
+import pkg/chronos/transports/common
+import pkg/chronos/streams/asyncstream
+import pkg/codex/streams
+import pkg/stew/byteutils
+
+import ./helpers
+
+asyncchecksuite "AsyncStreamWrapper":
+
+  let data = "0123456789012345678901234567890123456789"
+  let address = initTAddress("127.0.0.1:46001")
+
+  proc serveReadingClient(server: StreamServer,
+                          transp: StreamTransport) {.async.} =
+    var wstream = newAsyncStreamWriter(transp)
+    await wstream.write(data)
+    await wstream.finish()
+    await wstream.closeWait()
+    await transp.closeWait()
+    server.stop()
+    server.close()
+
+  proc serveWritingClient(buf: pointer, bufLen: int): auto =
+    return proc(server: StreamServer, transp: StreamTransport) {.async.} =
+      var rstream = newAsyncStreamReader(transp)
+      discard await rstream.readOnce(buf, bufLen)
+      await rstream.closeWait()
+      await transp.closeWait()
+      server.stop()
+      server.close()
+
+  test "Read all data":
+    var server = createStreamServer(address, serveReadingClient, {ReuseAddr})
+    server.start()
+
+    var transp = await connect(address)
+    var rstream = newAsyncStreamReader(transp)
+    var wrapper = AsyncStreamWrapper.new(reader = rstream)
+    var buf = newSeq[byte](data.len)
+    
+    let readLen = (await wrapper.readOnce(addr buf[0], buf.len))
+
+    await wrapper.closeImpl()
+    await transp.closeWait()
+    await server.join()
+
+    check rstream.closed()
+    check buf.len == readLen
+    check data.toBytes == buf
+
+  test "Read not all data":
+    var server = createStreamServer(address, serveReadingClient, {ReuseAddr})
+    server.start()
+
+    var transp = await connect(address)
+    var rstream = newAsyncStreamReader(transp)
+    var wrapper = AsyncStreamWrapper.new(reader = rstream)
+    var buf = newSeq[byte](data.len div 2)
+
+    let readLen = (await wrapper.readOnce(addr buf[0], buf.len))
+
+    await wrapper.closeImpl()
```

Comment:
We shouldn't call `closeImpl()` explicitly, it's a private field (hence `Impl`), calling `await wrapper.close()` is the proper way, which calls `closeImpl` internally. 

It's a bit unfortunate that `closeImpl` is marked as public, but I believe this is done because you can't have private inheritable `methods`.

Reply:
Thanks for the context. I also found the `Impl` suffix odd, but I thought it was a peculiar naming choice.